h5py/tests/test_dataset.py

```diff
@@ -217,6 +275,15 @@ def test_dtype_conflict(self):
         with self.assertRaises(TypeError):
             self.f.require_dataset('foo', (10, 3), 'S10')
 
+    def test_dtype_exact(self):
+        """ require_dataset with exactly dtype match """
+
+        dset = self.f.create_dataset('foo', (10, 3), 'f')
+        dset2 = self.f.require_dataset('foo', (10, 3), 'f', exact=True)
+        self.assertEqual(dset, dset2)
+        with self.assertRaises(TypeError):
+            self.f.require_dataset('foo', (10, 3), 'i', exact=True)
```

Comment:
This would also fail with `exact=False`, because `np.can_cast('f', 'i') == False`. To properly check `exact=True`, we should use a pair of dtypes which would be OK without it, e.g. create it with `'i8'` (int64) and require it with `'i4'` (int32).

Reply:
Yes, I know, but when `exact= False`, then it's another error type, I just want to cover these lines: https://github.com/h5py/h5py/blob/14e5542db20f1bbcdadd0a4d6a70fc05b191ceba/h5py/_hl/group.py#L225-L227

Reply:
It's a `TypeError` either way, though. And the point of tests is to check that things work as we expect, not just to cover lines of code. So I think it's a better test if it uses types which would be OK with `exact=False` but not with `exact=True`.